compio-driver/src/fd.rs

```diff
@@ -0,0 +1,152 @@
+use std::{
+    future::{poll_fn, ready, Future},
+    mem::ManuallyDrop,
+    panic::RefUnwindSafe,
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        Arc,
+    },
+    task::Poll,
+};
+
+use futures_util::{future::Either, task::AtomicWaker};
+
+use crate::{AsRawFd, OwnedFd, RawFd};
+
+#[derive(Debug)]
+struct Inner {
+    fd: OwnedFd,
+    waits: AtomicUsize,
+    waker: AtomicWaker,
+}
+
+impl RefUnwindSafe for Inner {}
+
+/// A shared fd. It is passed to the operations to make sure the fd won't be
+/// closed before the operations complete.
+#[derive(Debug, Clone)]
+pub struct SharedFd(Arc<Inner>);
+
+impl SharedFd {
+    /// Create the shared fd from an owned fd.
+    pub fn new(fd: impl Into<OwnedFd>) -> Self {
+        Self(Arc::new(Inner {
+            fd: fd.into(),
+            waits: AtomicUsize::new(0),
+            waker: AtomicWaker::new(),
+        }))
+    }
+
+    /// Try to take the inner owned fd.
+    pub fn try_owned(self) -> Result<OwnedFd, Self> {
+        let this = ManuallyDrop::new(self);
+        if let Some(fd) = Self::try_owned_inner(&this) {
+            Ok(fd)
+        } else {
+            Err(ManuallyDrop::into_inner(this))
+        }
+    }
+
+    fn try_owned_inner(this: &ManuallyDrop<Self>) -> Option<OwnedFd> {
```

Comment:
This by itself is not safe. You need to add `unsafe` and let caller to guarantee that if `Some` is returned, `Inner` should not be used again as `Arc` has already reset its counter.